include/sddf/sound/queue.h

```diff
@@ -0,0 +1,379 @@
+/*
+ * Copyright 2023, UNSW
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <stdint.h>
+#include <stddef.h>
+#include <stdbool.h>
+#include <sddf/sound/sound.h>
+#include <sddf/util/fence.h>
+
+#define SOUND_NUM_BUFFERS 512
+#define SOUND_PCM_BUFFER_SIZE 4096
+
+typedef enum {
+    SOUND_CMD_TAKE,
+    SOUND_CMD_PREPARE,
+    SOUND_CMD_RELEASE,
+    SOUND_CMD_START,
+    SOUND_CMD_STOP,
+} sound_cmd_code_t;
+
+typedef enum {
+    SOUND_S_OK = 0,
+    SOUND_S_BAD_MSG,
+    SOUND_S_NOT_SUPP,
+    SOUND_S_IO_ERR,
+    SOUND_S_BUSY,
+} sound_status_t;
+
+typedef struct sound_pcm_set_params {
+    uint8_t channels;
+    uint8_t format;
+    uint8_t rate;
+} sound_pcm_set_params_t;
+
+typedef struct sound_cmd {
+    sound_cmd_code_t code;
+    uint32_t cookie;
+    uint32_t stream_id;
+    union {
+        // Set on TAKE request
+        sound_pcm_set_params_t set_params;
+        // Set on all responses
+        sound_status_t status;
+    };
+} sound_cmd_t;
+
+typedef struct sound_pcm {
+    uint32_t cookie;
+    uint32_t stream_id;
+    uintptr_t addr;
+    unsigned int len;
+    // Only used in responses.
+    sound_status_t status;
+    uint32_t latency_bytes;
+} sound_pcm_t;
+
+// Eventually this could be moved into its own library
+typedef struct sound_queue_state {
+    uint32_t write_idx;
+    uint32_t read_idx;
+    uint32_t size;
```

Comment:
Should be moved into the component-local memory so that a nefarious client cannot mess with the size and cause an out of bounds error in the virtualiser.